### PR TITLE
Fix APIs that overwrite data in the shared cache.

### PR DESCRIPTION
**Pull request by hangshao0**
Store a new item in the top layer cache if JIT/GC wants to overwrite
an existing data in a lower layer cache.

issue #5480

Signed-off-by: Hang Shao <hangshao@ca.ibm.com>

### CODE REVIEW

File: runtime/shared_common/CacheMap.cpp [hunk 1]
```diff
@@ -3120,17 +3120,43 @@ SH_CacheMap::updateROMClassResource(J9VMThread* currentThread, const void* addre
...
+					memcpy(tmpItem, itemInCache, itemInCache->dataLen);
+					addResourceInTopLayer = true;
+					itemToUse = tmpItem;
+				}
 			}
 
-			const ShcItem *itemInCache = resourceDescriptor->wrapperToItem(resourceWrapper);
 			if (false == isUDATA) {
-				resourceDescriptor->updateDataInCache(itemInCache, updateAtOffset, data);
+				resourceDescriptor->updateDataInCache(itemToUse, updateAtOffset, data);
```
Comment (pshipton) [inline runtime/shared_common/CacheMap.cpp]: Shouldn't this only be done if addResourceInTopLayer is false?

File: runtime/shared_common/CacheMap.cpp [hunk 2]
```diff
@@ -4102,19 +4133,19 @@ SH_CacheMap::storeSharedData(J9VMThread* currentThread, const char* key, UDATA k
...
-		}
 	}
 
+_addData:
 	/* If data is NULL or datalen <= 0, mark the original item(s) stale, but don't store anything */
 	if ((data != NULL) && (data->length > 0) && ((data->address != NULL) || (data->flags & J9SHRDATA_ALLOCATE_ZEROD_MEMORY))) {
 		const J9UTF8* tokenKey = NULL;	
 
 		if (!dataNotIndexed) {
+			if (!(localSCM = getScopeManager(currentThread))) {
```
[the remark this answers is not in the document]
Comment (hangshao0) [reply runtime/shared_common/CacheMap.cpp]: Fixed

File: runtime/shared_common/CacheMap.cpp [hunk 3]
```diff
@@ -3120,17 +3120,43 @@ SH_CacheMap::updateROMClassResource(J9VMThread* currentThread, const void* addre
...
 			if (false == isAddressInCache((void*)updateAddress, data->length, false, true)) {
-				Trc_SHR_CM_updateROMClassResource_Exit7(currentThread, updateAddress, data->length);
-				result = J9SHR_RESOURCE_STORE_ERROR;
-				break;
+				/* We cannot overwrite the existing resource which is in a lower layer cache, instead we add a new resource. 
+				 * The old resources will be removed from the hashtable. */
+				Trc_SHR_Assert_True(isAddressInCache((void*)updateAddress, data->length, false, false));
+				tmpItem = (ShcItem*)j9mem_allocate_memory(itemInCache->dataLen, J9MEM_CATEGORY_CLASSES);
+				if (NULL == tmpItem) {
+					result = J9SHR_RESOURCE_STORE_ERROR;
```
[the remark this answers is not in the document]
Comment (hangshao0) [reply runtime/shared_common/CacheMap.cpp]: Tracepoint added

File: runtime/shared_common/CacheMap.cpp (hunk 1 above)
Comment (hangshao0) [reply runtime/shared_common/CacheMap.cpp]: When `addResourceInTopLayer` is false, itemToUse points to the `ShcItem` in the shared cache, so we are directly updating the data at `updateAtOffset` in the shared cache. 
When addResourceInTopLayer is true. `itemToUse` points to the local copy `tmpItem`, and we are updating the local copy at `updateAtOffset`. This local copy is then added to the top layer cache.

**Comment by hangshao0** [comment]
Doc change: https://github.com/eclipse/openj9-docs/issues/368

File: runtime/shared_common/shrinit.cpp [hunk 4]
```diff
@@ -255,6 +255,11 @@ J9SharedClassesHelpText J9SHAREDCLASSESHELPTEXT[] = {
 	{HELPTEXT_ADJUST_MAXAOT_EQUALS, J9NLS_SHRC_SHRINIT_HELPTEXT_ADJUST_MAXAOT_EQUALS, 0, 0},
 	{HELPTEXT_ADJUST_MINJITDATA_EQUALS, J9NLS_SHRC_SHRINIT_HELPTEXT_ADJUST_MINJIT_EQUALS, 0, 0},
 	{HELPTEXT_ADJUST_MAXJITDATA_EQUALS, J9NLS_SHRC_SHRINIT_HELPTEXT_ADJUST_MAXJIT_EQUALS, 0, 0},
+#if defined(J9VM_OPT_MULTI_LAYER_SHARED_CLASS_CACHE)
```
Comment (pshipton) [inline runtime/shared_common/shrinit.cpp]: I think moving the options to the help should wait until the remaining layer support and testing is delivered. We'll target the 0.18 release.

Comment (hangshao0) [reply runtime/shared_common/shrinit.cpp]: Removed this change.

**Comment by pshipton** [comment]
jenkins test sanity win jdk8

**Pull request merged**